### ISSUE TITLE
Tooltip mode

### ISSUE THREAD

**Issue by birkenfeld** [rocky/emacs-dbgr#38]
GUD has a tooltip mode where it shows symbol values as a tooltip on mouse hover. It would be nice to have the same in realgud (e.g. in the Python debugger mode).

**Comment by rocky** [comment]
I either don't understand what you are talking about or don't experience this behavior. (Or both)

There is much about Emacs that I don't use or understand that other may take for granted. So it would be very helpful to be every precise ratther than general. 

When I run realgud such as realgud:trepan2, I see in the toobar menu "Set Breakpoint", "Restart", "Continue" and some icons for "finish", "Step out"  ,"step"  and ending with "Up Stack". When mouse over these, I see tool tips. For example hovering over "Set Breakpoint" the tooltip is "Set Breakpoint at the current line." 

Is this what you get too?  Or you are trying something different? Of course, when I run tooltip-mode and it is disabled, those tips aren't shown. Just in case you haven't checked this, see if tooltip-mode is enabled.

**Comment by birkenfeld** [comment]
I agree, I was a bit imprecise. I'll try to make a screenshot of the feature I was talking about.

**Comment by birkenfeld** [comment]
Here it is: the tooltip shows the value of the symbol under the mouse. It is implemented in `gud-tooltip-mode`.

![tooltip](https://cloud.githubusercontent.com/assets/144359/6208322/35bc5b28-b5b7-11e4-8147-9e2d02c7d7e0.png)

**Comment by rocky** [comment]
That's pretty cool. I wonder how it works? That code that starts define-minor-mode and the function gud-tool-tips in gud.el looks pretty impenetrable.

**Comment by birkenfeld** [comment]
I'm afraid I couldn't answer this, being mostly an Emacs user, not hacker :)

**Comment by rocky** [comment]
I've now traced through the code and understand what's going on at least for the portion that is retrieving the value. I may do this in two steps where the first step you hit a key to eval. And even here I'll probably need to block out a large portion of coding time to get this done, even though translating the code seems pretty straightforward. (The same set of concepts exists in both gdb and realgud)

**Comment by rocky** [comment]
With commit [015358a66410cb3d7693755f1f4df0c6e91cd501] if you are in a source-code buffer in short-key mode and you hit mouse-2 the expression under the mouse or the region that is selected there will get evaluated and displayed as a tooltip. 

Perhaps not exactly the same behavior. The gud code has a lot of frills for setting things like whether to show the tooltip in the echo area or not. And there are various bugs like sometimes I see this messing up highlighting processing. But this is a start. Over time it things will get ironed out more.

**Comment by dgutov** [comment]
I suppose showing it automatically isn't safe anyway. At least in Ruby, displaying a value is a potentially destructive operation, and it's hard for the debugger interface to know in advance.

**Comment by rocky** [comment]
In contrast to gud.el you click to get an expression evaluated.

**Comment by dgutov** [comment]
True. I suppose I'm just wondering why this issue is still open.

**Comment by rocky** [comment]
@birkenfeld there's been time to changes to this settle down. Due to the state-changing behavior of eval, in contrast to what's there with gdb-mi the behavior for getting results I think has to be a user requesting evaluation rather than it automatically (and even perhaps inadvertently) getting run. If you have an objection to closing this issue, please let me kno.

**Comment by dgutov** [comment]
@rocky Thinking more about this, maybe this can be a slightly different feature. Instead of evaluating expressions, it can try to limit itself to entities and representations each specific debugger considers safe-ish.

`gdb` can do what's displayed on the earlier screenshot.

`rdebug` family of Ruby debuggers has a `display` command that calls `#to_s` but not `#inspect`. It's overridable as well, so not 100% safe either, but `ruby-debug-ide`, for instance, seems to depend on it:

https://www.jetbrains.com/ruby/features/screenshots/debug/Rails_debugger.png
https://www.jetbrains.com/ruby/features/screenshots/debug/Rails_debugger2.png

It wouldn't be a good debugger if the local variables display routinely screwed with evaluation, right?

**Comment by rocky** [comment]
I'm of two minds. Yes, calling either `#to_s` or even `#inspect` in Ruby or `str()` and `repr()` in Python is, in general less state changing than calling `eval()` in Ruby;  (In Python `eval()` doesn't work on statements like it does in Python; instead `exec()` does that.)

But still, in either language state can get changed. 

In https://www.jetbrains.com/ruby/features/screenshots/debug/Rails_debugger.png   `faq.question` could be a function call that has side effects.  And although it would be safe to run `"faq.question".to_s`, that wouldn't be very helpful. 

Yes, the ameliorate things, it would be nice to have  a buffer containing local variable names their values.  Or maybe there might be variable to control whether to allow `#eval`, `#inspect`, or `#to_s`   getting run without confirmation.  

In the past, I've gotten problem reports as a result of doing such things. It's an instance of a [Heisenbug](http://en.wikipedia.org/wiki/Heisenbug).

**Comment by dgutov** [comment]
> faq.question could be a function call that has side effects

Note that I'm not talking about the `Watches` pane: any watch (or other arbitrary expression evaluation) should be requested explicitly by the user.

Only the `Variables` pane. We can also ask the debugger for the list of the current local variables, so we don't mistake a method call with implicit receiver for a local var. Being able to entirely disable this via user option would be useful too, of course.

**Comment by kootenpv** [comment]
I really like this idea... could it be done? This feature would really make emacs more future proof...

**Comment by rocky** [comment]
@kootenpv no one has worked on a `Variables` pane. Feel free to work on it.
